File: packages/app/src/services/software/definitions/keepass.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'keepass';
	public name = 'KeePass';
	public category = [SoftwareCategory.Security];
	public downloadName = 'KeePass-Setup.exe';
	public isArchive = false;
	public shouldCacheUrl = false;
	public requiresAdmin = false;
	public deprecated = false;
	public icon = 'keepass.png';
	public homepage = 'https://keepass.info';

	public async resolveDownloadUrl() {
		const downloadPageUrl = await this.getDownloadPageUrl();
		if (!downloadPageUrl) {
			return err(DownloadUrlResolveError.Generic);
		}

		const res = await fetch(downloadPageUrl);
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/<a href="(.*)" rel="nofollow">/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}

	private async getDownloadPageUrl() {
		const res = await fetch('https://keepass.info/download.html');
		if (!res.ok) {
			return null;
		}

		const html  = await res.text();
		const match = html.match(/https:\/\/sourceforge\.net\/projects\/keepass\/files\/(KeePass%202\.x\/\d+\.\d+(\.\d+)?\/KeePass-\d+\.\d+(\.\d+)?-Setup\.exe)\/download/);
		if (!match) {
			return null;
		}

		return `https://sourceforge.net/settings/mirror_choices?projectname=keepass&filename=${match[1]}&dialog=true`;
	}
}

File: src/lib/software/keepass.py
from re import compile
from PySide6.QtCore import Slot, QUrl
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class KeePass(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'keepass'
        self.name = 'KeePass'
        self.category = [SoftwareCategory.Security]
        self.download_name = 'KeePass-Setup.exe'
        self.icon = 'keepass.png'
        self.homepage = 'https://keepass.info'

        self._initial_url = QUrl('https://keepass.info/download.html')

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        html = reply.readAll().data().decode()

        if reply.url() == self._initial_url:
            download_file_name_pattern = compile(r'https://sourceforge\.net/projects/keepass/files/(KeePass%202\.x/\d+\.\d+(\.\d+)?/KeePass-\d+\.\d+(\.\d+)?-Setup\.exe)/download')
            match = download_file_name_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                req = QNetworkRequest(f'https://sourceforge.net/settings/mirror_choices?projectname=keepass&filename={match.group(1)}&dialog=true')
                self.manager.get(req)
        else:
            download_url_pattern = compile(r'<a href="(.*)" rel="nofollow">')
            match = download_url_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self._initial_url))
